packages/yarnpkg-core/sources/nodeUtils.ts

```diff
@@ -1,6 +1,69 @@
-import Module from 'module';
+import Module         from 'module';
+
+import * as miscUtils from './miscUtils';
 
 export function builtinModules(): Set<string> {
   // @ts-expect-error
   return new Set(Module.builtinModules || Object.keys(process.binding(`natives`)));
 }
+
+function getLibc() {
+  const report: any = process.report?.getReport() ?? {};
+  const sharedObjects: Array<string> = report.sharedObjects ?? [];
+
+  const libcRegExp = /\/(?:(ld-linux)|(ld-musl))-/;
+
+  return miscUtils.mapAndFind(sharedObjects, entry => {
+    const match = entry.match(libcRegExp);
+    if (!match)
+      return miscUtils.mapAndFind.skip;
+
+    if (match[1])
+      return `glibc`;
+    if (match[2])
+      return `musl`;
+
+    throw new Error(`Assertion failed: Expected the libc variant to have been detected`);
+  }) ?? null;
+}
```

Comment:
@lovell Does that sound a reasonable heuristic?

Reply:
Based on the existing [detect-libc filename-based logic](https://github.com/lovell/detect-libc/blob/88df1a5950bf3cd9bffa1e0137ab6471c4546118/lib/detect-libc.js#L61-L72), it might be worth matching on `linux-gnu` and `libc.musl` too. I'm currently updating the integration test matrix and will let you know if I find any other relevant filename patterns.